File: server/internal/reminders/reminder.go
package reminders

import (
	"time"

	"gorm.io/gorm"
)

// RecurrenceType defines the supported recurrence patterns
type RecurrenceType string

const (
	RecurrenceNone    RecurrenceType = "none"    // One-time reminder
	RecurrenceDaily   RecurrenceType = "daily"   // Every N days
	RecurrenceWeekly  RecurrenceType = "weekly"  // Every N weeks, on specific days
	RecurrenceMonthly RecurrenceType = "monthly" // Every N months, on specific day
	RecurrenceYearly  RecurrenceType = "yearly"  // Every N years, on specific month/day
)

// Reminder represents a user reminder (one-time or recurring)
type Reminder struct {
	ID    uint   `json:"id" gorm:"primaryKey"`
	Title string `json:"title" gorm:"not null"`
	// Detail is the body text of the reminder
	Detail string `json:"detail"`

	// Timezone stores the IANA timezone used for recurrence calculations
	// (e.g. "America/New_York").
	Timezone string `json:"timezone" gorm:"type:varchar(100);not null;default:'UTC'"`

	// RemindAt is the initial/one-time reminder datetime (user-specified)
	RemindAt time.Time `json:"remind_at" gorm:"not null"`

	// RemindHour is the hour (0-23) at which the reminder should fire.
	// For recurring reminders this determines the daily trigger time.
	RemindHour int `json:"remind_hour" gorm:"not null;default:9"`

	// Recurrence fields
	RecurrenceType     RecurrenceType `json:"recurrence_type" gorm:"type:varchar(20);not null;default:'none'"`
	RecurrenceInterval int            `json:"recurrence_interval" gorm:"not null;default:1"` // every N periods

	// RecurrenceDaysOfWeek: comma-separated days for weekly recurrence.
	// 0=Sun, 1=Mon, 2=Tue, 3=Wed, 4=Thu, 5=Fri, 6=Sat
	// e.g. "1,3,5" means Mon, Wed, Fri
	RecurrenceDaysOfWeek string `json:"recurrence_days_of_week" gorm:"type:varchar(20)"`

	// RecurrenceDayOfMonth: day of month for monthly recurrence (1-31).
	// 0 means last day of month.
	RecurrenceDayOfMonth int `json:"recurrence_day_of_month" gorm:"default:0"`

	// RecurrenceMonthOfYear: month for yearly recurrence (1-12)
	RecurrenceMonthOfYear int `json:"recurrence_month_of_year" gorm:"default:0"`

	// RecurrenceEndDate: optional end date for recurring reminders (nullable)
	RecurrenceEndDate *time.Time `json:"recurrence_end_date"`

	// NextRemindAt is the pre-computed next trigger time. Indexed for efficient querying.
	// The background service only checks: WHERE is_active = true AND next_remind_at <= NOW()
	NextRemindAt time.Time `json:"next_remind_at" gorm:"not null;index"`

	// LastRemindedAt tracks when the reminder was last triggered
	LastRemindedAt *time.Time `json:"last_reminded_at"`

	// IsActive indicates whether the reminder is still active
	IsActive bool `json:"is_active" gorm:"not null;default:true;index"`

	// CompletedAt is set when a one-time reminder fires or a recurring reminder is manually completed
	CompletedAt *time.Time `json:"completed_at"`

	UserID    uint           `json:"user_id" gorm:"not null;index"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `json:"-" gorm:"index"`
}

File: ios/jiceotapp/Models/Reminder.swift
//
//  Reminder.swift
//  JiceotApp
//
//  Created for Reminders feature.
//

import Foundation

struct Reminder: Codable, Identifiable {
    let id: Int
    let title: String
    let detail: String
    let timezone: String
    let remindAt: String
    let remindHour: Int
    let recurrenceType: String
    let recurrenceInterval: Int
    let recurrenceDaysOfWeek: String
    let recurrenceDayOfMonth: Int
    let recurrenceMonthOfYear: Int
    let recurrenceEndDate: String?
    let nextRemindAt: String
    let lastRemindedAt: String?
    let isActive: Bool
    let completedAt: String?
    let userId: Int
    let createdAt: String
    let updatedAt: String
    
    enum CodingKeys: String, CodingKey {
        case id
        case title
        case detail
        case timezone
        case remindAt = "remind_at"
        case remindHour = "remind_hour"
        case recurrenceType = "recurrence_type"
        case recurrenceInterval = "recurrence_interval"
        case recurrenceDaysOfWeek = "recurrence_days_of_week"
        case recurrenceDayOfMonth = "recurrence_day_of_month"
        case recurrenceMonthOfYear = "recurrence_month_of_year"
        case recurrenceEndDate = "recurrence_end_date"
        case nextRemindAt = "next_remind_at"
        case lastRemindedAt = "last_reminded_at"
        case isActive = "is_active"
        case completedAt = "completed_at"
        case userId = "user_id"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
    
    var recurrenceDescription: String {
        let interval = recurrenceInterval > 0 ? recurrenceInterval : 1
        switch recurrenceType {
        case "none":
            return "One-time"
        case "daily":
            return interval == 1 ? "Daily" : "Every \(interval) days"
        case "weekly":
            let dayNames = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"]
            let days = recurrenceDaysOfWeek.split(separator: ",")
                .compactMap { Int($0.trimmingCharacters(in: .whitespaces)) }
                .compactMap { $0 >= 0 && $0 <= 6 ? dayNames[$0] : nil }
                .joined(separator: ", ")
            let prefix = interval == 1 ? "Weekly" : "Every \(interval) weeks"
            return days.isEmpty ? prefix : "\(prefix) on \(days)"
        case "monthly":
            let day = recurrenceDayOfMonth == 0 ? "last day" : "day \(recurrenceDayOfMonth)"
            return interval == 1 ? "Monthly on \(day)" : "Every \(interval) months on \(day)"
        case "yearly":
            let formatter = DateFormatter()
            formatter.dateFormat = "MMM"
            let monthName = recurrenceMonthOfYear >= 1 && recurrenceMonthOfYear <= 12
                ? formatter.monthSymbols[recurrenceMonthOfYear - 1]
                : ""
            let day = recurrenceDayOfMonth == 0 ? "last day" : "\(recurrenceDayOfMonth)"
            return interval == 1
                ? "Yearly on \(monthName) \(day)"
                : "Every \(interval) years on \(monthName) \(day)"
        default:
            return "Unknown"
        }
    }
}

struct CreateReminderRequest: Codable {
    let title: String
    let detail: String?
    let timezone: String
    let remindAt: String
    let remindHour: Int
    let recurrenceType: String
    let recurrenceInterval: Int?
    let recurrenceDaysOfWeek: String?
    let recurrenceDayOfMonth: Int?
    let recurrenceMonthOfYear: Int?
    let recurrenceEndDate: String?
    
    enum CodingKeys: String, CodingKey {
        case title
        case detail
        case timezone
        case remindAt = "remind_at"
        case remindHour = "remind_hour"
        case recurrenceType = "recurrence_type"
        case recurrenceInterval = "recurrence_interval"
        case recurrenceDaysOfWeek = "recurrence_days_of_week"
        case recurrenceDayOfMonth = "recurrence_day_of_month"
        case recurrenceMonthOfYear = "recurrence_month_of_year"
        case recurrenceEndDate = "recurrence_end_date"
    }
}

struct UpdateReminderRequest: Codable {
    let title: String
    let detail: String?
    let timezone: String
    let remindAt: String
    let remindHour: Int
    let recurrenceType: String
    let recurrenceInterval: Int?
    let recurrenceDaysOfWeek: String?
    let recurrenceDayOfMonth: Int?
    let recurrenceMonthOfYear: Int?
    let recurrenceEndDate: String?
    let isActive: Bool
    
    enum CodingKeys: String, CodingKey {
        case title
        case detail
        case timezone
        case remindAt = "remind_at"
        case remindHour = "remind_hour"
        case recurrenceType = "recurrence_type"
        case recurrenceInterval = "recurrence_interval"
        case recurrenceDaysOfWeek = "recurrence_days_of_week"
        case recurrenceDayOfMonth = "recurrence_day_of_month"
        case recurrenceMonthOfYear = "recurrence_month_of_year"
        case recurrenceEndDate = "recurrence_end_date"
        case isActive = "is_active"
    }
}

struct ReminderListResponse: Codable {
    let reminders: [Reminder]
    let total: Int
}
